homeassistant/components/tplink/config_flow.py

```diff
@@ -137,16 +313,109 @@ def _async_create_entry_from_device(self, device: SmartDevice) -> FlowResult:
+        )
+        return self._discovered_device
+
     async def _async_try_connect(
-        self, host: str, raise_on_progress: bool = True
+        self,
+        discovered_device: SmartDevice,
+        credentials: Optional[Credentials],
     ) -> SmartDevice:
         """Try to connect."""
-        self._async_abort_entries_match({CONF_HOST: host})
-        device: SmartDevice = await Discover.discover_single(host)
+        self._async_abort_entries_match({CONF_HOST: discovered_device.host})
+
+        config = discovered_device.config
+        if credentials:
+            config.credentials = credentials
+        config.timeout = CONNECT_TIMEOUT
+        if config.uses_http:
+            config.http_client = create_async_httpx_client(self.hass, verify_ssl=False)
+
+        self._discovered_device = await SmartDevice.connect(config=config)
         await self.async_set_unique_id(
-            dr.format_mac(device.mac), raise_on_progress=raise_on_progress
+            dr.format_mac(self._discovered_device.mac),
+            raise_on_progress=False,
+        )
+        return self._discovered_device
+
+    async def async_step_reauth(self, entry_data: Mapping[str, Any]) -> FlowResult:
+        """Handle reauth upon an API authentication error."""
+        self.reauth_entry = self.hass.config_entries.async_get_entry(
+            self.context["entry_id"]
+        )
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Dialog that informs the user that reauth is required."""
+        errors = {}
+        assert self.reauth_entry is not None
+
+        if user_input:
+            host = self.reauth_entry.data[CONF_HOST]
+            credentials = Credentials(
+                user_input[CONF_USERNAME], user_input[CONF_PASSWORD]
+            )
+            try:
+                await self._async_try_discover_and_update(
+                    host, credentials=credentials, raise_on_progress=True
+                )
+            except AuthenticationException:
+                errors["base"] = "invalid_auth"
+            except SmartDeviceException:
+                errors["base"] = "unknown"
+            else:
+                await set_credentials(
+                    self.hass, user_input[CONF_USERNAME], user_input[CONF_PASSWORD]
+                )
+                await self.hass.config_entries.async_reload(self.reauth_entry.entry_id)
+                async_call_later(
+                    self.hass,
+                    REAUTH_REFLOW_CALL_LATER_SECS,
+                    self._async_reload_requires_auth_entries,
+                )
+                return self.async_abort(reason="reauth_successful")
+
+        # Old config entries will not have these values.
+        alias = self.reauth_entry.data.get(CONF_ALIAS) or "[alias]"
+        model = self.reauth_entry.data.get(CONF_MODEL) or "[model]"
```

Comment:
This is a bit confusing.  Maybe `unknown`?

Reply:
Changed to unknown.  This is a very unlikely scenario now as `async_setup_entry` will update these values if they don't exist.